juju/juju
# Local credentials tolerate per-cloud malformed credentials.

## Description of change

Malformed credential.yaml files produce unhelpful errors and prevent the user
from moving forward despite the command not requiring cloud credential that is malformed.

For example `juju bootstrap --clouds`, `juju remove-credential`, `juju list-credentials` etc that depend on a cloud credential from the local credential store will fail regardless of if they need the credential.

The change allows the local file credential store via CredentialForCloud to access credentials without requiring the whole credentials.yaml file to be validate. Instead validation is performed lazily when the
particular credential is accessed.

Both `juju bootstrap --clouds` and `juju list-credentials` treat individual cloud credential entries independently and warn instead of returning an error.

## QA steps

Alter an entry in credentials.yaml and change the field name `auth-type` to `auth_type` to force validation to fail.

Run `juju bootstrap --clouds` and it should WARN with the credential error, but should still list available clouds to bootstrap.

Run `juju list-credentials` and it should WARN with the credential error, but should still list available clouds and their information (except the invalid ones).

Run `juju bootstrap localhost` and it should deploy successfully and mutate the `credentials.yaml` file leaving the malformed entry.

## Documentation changes

Credentials that don't map to a cloud name no longer list in `juju list-credentials`

## Bug reference

[https://bugs.launchpad.net/juju/+bug/1616523](https://bugs.launchpad.net/juju/+bug/1616523)
